app/lib/delivery_failure_tracker.rb

```diff
@@ -3,47 +3,53 @@
 class DeliveryFailureTracker
   FAILURE_DAYS_THRESHOLD = 7
 
-  def initialize(inbox_url)
-    @inbox_url = inbox_url
+  def initialize(url_or_host)
+    @host = (url_or_host.start_with?('https://') || url_or_host.start_with?('http://')) ? Addressable::URI.parse(url_or_host).normalized_host : url_or_host
   end
 
   def track_failure!
     Redis.current.sadd(exhausted_deliveries_key, today)
-    Redis.current.sadd('unavailable_inboxes', @inbox_url) if reached_failure_threshold?
+    UnavailableDomain.create(domain: @host) if reached_failure_threshold?
   end
 
   def track_success!
     Redis.current.del(exhausted_deliveries_key)
-    Redis.current.srem('unavailable_inboxes', @inbox_url)
+    UnavailableDomain.find_by(domain: @host)&.destroy
   end
 
   def days
     Redis.current.scard(exhausted_deliveries_key) || 0
   end
 
+  def available?
+    !UnavailableDomain.where(domain: @host).any?
+  end
+
+  alias reset! track_success!
+
   class << self
-    def filter(arr)
-      arr.reject(&method(:unavailable?))
-    end
+    def without_unavailable(urls)
+      unavailable_domains_map = Rails.cache.fetch('unavailable_domains') { UnavailableDomain.pluck(:domain).each_with_object({}) { |domain, hash| hash[domain] = true } }
```

Comment:
The number of unavailable domains could grow quite a bit, I'm not sure loading them all is safe…

Maybe caching a bloom filter instead, and then querying the database for domains that are possibly unavailable only would make more sense? Although rebuilding the bloom filter each time a domain is made available again would be costly…

Reply:
You're not wrong but at the same time the number of distinct hostnames will be smaller than the current number of distinct inbox URLs. I think bloom filters are a nice idea but maybe something one could do down the line when this becomes practical bottleneck.

Reply:
That is true, but although I'm not sure how Redis handles large sets, I'd think it handles it better than just creating a big Ruby hash for it… honestly I'm not sure what's best here, but loading the whole thing really doesn't sound right.